File: apps/web/src/features/lab-analysis/schemas.ts
import { z } from "zod";

const optionalNumber = z.preprocess((value) => {
  if (value === "" || value === null || Number.isNaN(value)) return undefined;
  return value;
}, z.coerce.number().optional());


export const labEntryFormSchema = z.object({
  patientContext: z.object({
    age: z.preprocess((value) => {
      if (value === "" || value === null || Number.isNaN(value)) return undefined;
      return value;
    }, z.coerce.number().min(0).max(130).optional()),
    sex: z.enum(["female", "male", "other", "unknown"]).default("unknown")
  }),
  results: z.array(
    z.object({
      testName: z.string().min(1, "Test name is required"),
      value: z.coerce.number({ invalid_type_error: "Value must be a number" }),
      unit: z.string().optional(),
      referenceRange: z.object({
        low: optionalNumber,
        high: optionalNumber
      }),
      notes: z.string().optional()
    })
  ).min(1)
});

export type LabEntryFormValues = z.infer<typeof labEntryFormSchema>;

File: apps/api/src/modules/lab_analysis/schemas.py
from typing import Literal
from pydantic import BaseModel, Field, field_validator

LabStatus = Literal["normal", "low", "high", "borderline", "unknown"]
LabSeverity = Literal["none", "mild", "moderate", "critical", "unknown"]


class ReferenceRange(BaseModel):
    low: float | None = None
    high: float | None = None
    text: str | None = None


class PatientContext(BaseModel):
    age: int | None = Field(default=None, ge=0, le=130)
    sex: Literal["male", "female", "other", "unknown"] | None = "unknown"
    pregnant: bool | None = None


class LabInput(BaseModel):
    testName: str = Field(min_length=1, max_length=120)
    value: float
    unit: str | None = Field(default=None, max_length=40)
    referenceRange: ReferenceRange | None = None
    notes: str | None = Field(default=None, max_length=500)
    source: Literal["manual", "pdf"] | None = "manual"

    @field_validator("testName")
    @classmethod
    def strip_test_name(cls, value: str) -> str:
        return value.strip()


class AnalyzeRequest(BaseModel):
    patientContext: PatientContext | None = None
    results: list[LabInput] = Field(min_length=1, max_length=50)


class ClassifiedLabResult(BaseModel):
    testName: str
    normalizedName: str
    value: float
    unit: str | None = None
    referenceRange: ReferenceRange | None = None
    status: LabStatus
    severity: LabSeverity
    source: Literal["manual", "pdf"] | None = "manual"


class CombinationFlag(BaseModel):
    code: str
    title: str
    explanation: str
    severity: Literal["info", "follow_up", "urgent"] = "follow_up"


class LabExplanation(BaseModel):
    testName: str
    normalizedName: str
    value: float
    unit: str | None = None
    referenceRange: ReferenceRange | None = None
    status: LabStatus
    severity: LabSeverity
    plainLanguageExplanation: str
    possibleGeneralCauses: list[str]
    followUpQuestions: list[str]
    shouldDiscussWithClinician: bool
    disclaimer: str = "This explanation is educational and is not a medical diagnosis."


class AnalysisSummary(BaseModel):
    totalResults: int
    normalCount: int
    abnormalCount: int
    followUpRecommended: bool
    overallPlainLanguageSummary: str
    importantNotes: list[str]


class AnalyzeResponse(BaseModel):
    summary: AnalysisSummary
    results: list[LabExplanation]
    combinationFlags: list[CombinationFlag]
